purldb-toolkit/src/purldb_toolkit/purlcli.py

```diff
@@ -623,64 +703,86 @@ def get_versions(purls, output, file):
     json.dump(purl_versions, output, indent=4)
 
 
-# construct_headers() has not yet been implemented for this `versions` command
-# -- or for the `validate` command.
-def list_versions(purls, output, file, command_name):
+def get_versions_details(purls, output, file, unique, command_name):
     """
     Return a list of dictionaries containing version-related data for each PURL
     in the `purls` input list.  `check_versions_purl()` will print an error
     message to the console (also displayed in the JSON output) when necessary.
     """
-    purl_versions = []
-    for purl in purls:
-        purl_data = {}
-        purl_data["purl"] = purl
-        purl_data["versions"] = []
+    versions_details = {}
+    versions_details["headers"] = []
+    versions_details["packages"] = []
+
+    versions_warnings = {}
+
+    input_purls, normalized_purls = normalize_purls(purls, unique)
 
+    clear_log_file()
+
+    for purl in input_purls:
         purl = purl.strip()
         if not purl:
             continue
 
-        versions_purl = check_versions_purl(purl)
-
-        if command_name == "versions" and versions_purl == "not_valid":
-            print(f"'{purl}' not valid")
-            continue
+        purl_data = {}
+        purl_data["purl"] = purl
+        purl_data["versions"] = []
 
-        if command_name == "versions" and versions_purl == "valid_but_not_supported":
-            print(f"'{purl}' not supported with `versions` command")
-            continue
+        versions_purl = check_versions_purl(purl)
 
-        if command_name == "versions" and versions_purl == "not_in_upstream_repo":
-            print(f"'{purl}' does not exist in the upstream repo")
+        if command_name == "versions" and versions_purl:
+            versions_warnings[purl] = versions_purl
             continue
 
-        for package_version_object in list(versions(purl)):
+        for package_version in list(versions(purl)):
             purl_version_data = {}
-            purl_version = package_version_object.to_dict()["value"]
-            nested_purl = purl + "@" + f"{purl_version}"
+            purl_version = package_version.to_dict()["value"]
+
+            # We use `versions()` from fetchcode/package_versions.py, which
+            # keeps the version (if any) of the input PURL in its output, so
+            # "pkg:pypi/fetchcode@0.3.0" is returned as
+            # "pkg:pypi/fetchcode@0.3.0@0.1.0", "pkg:pypi/fetchcode@0.3.0@0.2.0"
+            # etc.  Thus, we remove any string starting with `@` first.
+            raw_purl = purl = re.split("[@,]+", purl)[0]
+            nested_purl = raw_purl + "@" + f"{purl_version}"
 
             purl_version_data["purl"] = nested_purl
             purl_version_data["version"] = f"{purl_version}"
             purl_version_data["release_date"] = (
-                f'{package_version_object.to_dict()["release_date"]}'
+                f'{package_version.to_dict()["release_date"]}'
```

Comment:
```suggestion
                f'{package_version.release_date}'
```

Reply:
@JonoYang Very interesting.  It took me a while to figure out why we could replace `package_version.to_dict()["value"]` with a dot approach -- `package_version.value`.  I saw nothing in `purlcli.py` itself that gave me a hint (`from fetchcode.package_versions import SUPPORTED_ECOSYSTEMS, versions` was eventually an indirect hint).  

I finally found that in fetchcode's `package_versions.py` we have a `@dataclasses.dataclass` decorator that defines the `PackageVersion` class -- a class whose name I'd previously noticed in a console error message when I tried modifying the name after the dot, wondering how this worked.  

I see the `value` and `release_date` attributes defined in that class, but how is it that I can use that class in `purlcli.py` without importing the class (and we don't, at least not by name)?

Reply:
With the change, 4 tests now fail ....

Reply:
@johnmhoran 

> I see the value and release_date attributes defined in that class, but how is it that I can use that class in purlcli.py without importing the class (and we don't, at least not by name)?

At the top of `purlcli.py`, you have the import statement `from fetchcode.package_versions import SUPPORTED_ECOSYSTEMS, versions` . Since `PackageVersion` is defined in `fetchcode.package_versions`, that is also imported

Reply:
Done.  The test failures were caused when some, but not all, of the generated `release_date` values used a slightly different format, e.g.,

```
'2013-12-11T00:27:10.097000+00:00'
vs.
'2013-12-11 00:27:10.097000+00:00'

```